include/io/mdarray.hpp

```diff
@@ -0,0 +1,274 @@
+#ifndef NGEN_IO_MDARRAY_HPP
+#define NGEN_IO_MDARRAY_HPP
+
+#include <initializer_list>
+#include <stdexcept>
+#include <vector>
+
+namespace io {
+
+template<typename T>
+class mdarray
+{
+  public:
+    using value_type      = T;
+    using container_type  = std::vector<value_type>;
+    using ilist           = std::vector<std::size_t>;
+    using size_type       = typename container_type::size_type;
+    using difference_type = typename container_type::difference_type;
+    using reference       = typename container_type::reference;
+    using const_reference = typename container_type::const_reference;
+    using pointer         = typename container_type::pointer;
+    using const_pointer   = typename container_type::const_pointer;
+
+    struct iterator;
+    friend iterator;
+
+    // Deleted default constructor. mdarray must be given a rank at least.
+    mdarray() = delete;
+
+    mdarray(size_type rank) noexcept
+        : m_shape(rank)
+        , m_data() {};
+
+    mdarray(ilist dsizes)
+        : m_shape(dsizes.begin(), dsizes.end())
+        , m_data(this->max_size()) {};
+
+    /**
+     * Retrieve a reference to the value at the given index.
+     *
+     * @example
+     * mdarray<int> x(2); // rank 2
+     * x.at({0, 0});      // get the value at (0, 0) in a 2D array.
+     * x.at({1, 1}) = 3;  // set the value at (1, 1) in a 2D array.
+     *
+     * @param n Index list
+     * @return reference 
+     */
+    reference at(const ilist& n)
```

Comment:
I'm bothered by this using `std::vector` as the argument type for indexing, because it means that most every call site is going to have to construct a temporary to pass, including heap allocation and de-allocation.

I don't immediately have a concrete suggestion for a complete replacement, but one possibility would be taking `std::initializer_list<size_t>`

I *think* that if we had `std::span` (alas, C++20) or an analogue available, we could just take that, and both `vector`s and braced initializer lists would be accepted, with no notable overhead.

I plan on proposing the import of `boost::span` into our codebase to better support passing data between models and layers of the framework, so I'm glad to see there's another prospective use case.

Reply:
I think, when this member is used, the vector argument will almost always be constructed before calling. The tests don't necessarily reflect that, but the handlers (for example) do.

I agree that a `std::span` impl would be better served here, but I'm not sure that a `std::initializer_list<size_t>` would work well here (since this prevents a vector from being passed in, I think?)

Reply:
See also #576 , since this is still open.